Add fuzzer

josharian opened the pull request:
This is an updated version of https://github.com/pion/sdp/pull/12 by @maxhawkins.

josharian commented:
This now also includes a test and fix for #13, which I rediscovered while running this fuzzer.

at-wat commented on fuzz.go:
```diff
@@ -0,0 +1,31 @@
+	// Check that we can marshal anything we unmarshalled.
+	_, err := sd.Marshal()
+	if err != nil {
+		panic("failed to marshal")
```
https://travis-ci.org/github/pion/sdp/jobs/690765726#L270
```
$ bash .github/lint-disallowed-functions-in-library.sh
/home/travis/gopath/src/github.com/pion/sdp/.github/../fuzz.go:		panic("failed to marshal")
panic( may only be used in example code
```
please add `// nolint`

josharian replied:
Done. If we add many more fuzz functions we'll probably want to make an exception for any file guarded by `// +build gofuzz`.

at-wat commented on unmarshal.go:
```diff
@@ -601,9 +601,9 @@ func unmarshalConnectionInformation(value string) (*ConnectionInformation, error
 	}
 
-	var connAddr *Address
+	connAddr := new(Address)
```
This seems adding whitespace at the end of `c=` line like `v=0\no=0 0 0 IN IP4 0\ns=0\nc=IN IP4 \nt=0 0\n`.
It would be nice to have a roundtrip test.

josharian replied:
This is not due to this change; it is because of how ConnectionInformation.String is generated. I've added a commit to this PR to fix it, though.

josharian commented:
I've added my name to the readme, although I'm not happy about it. I'll raise that on slack.

I've also added some performance optimizations. There's plenty more low hanging fruit after this, but it is a start.

josharian commented:
The build failure is because it claims that "marshalling" is the wrong spelling.

https://stackoverflow.com/questions/4434590/marshall-or-marshal-marshalling-or-marshaling

The wikipedia entry contains both spellings: https://en.wikipedia.org/wiki/Marshalling_(computer_science)

I will change it, but under protest.

josharian commented:
Travis failed this PR, saying:

> it failed 'Separate subject from body with a blank line'

But I don't see that any of my commit messages match this description. (And, unhelpfully, it doesn't appear to tell me *which* commit message is the problem.) I extracted the test it is doing  (`awk 'NR == 2 {print $1;}' | wc -c != 1` and ran it on each commit in turn, and they all passed.

Any hints?

I've now spent more time trying to fix lint complaints (spelling, commit messages, where to write my name) than actually working on the code.

josharian commented:
OK, the CI finally passes on this PR.

Sean-Der commented:
@josharian amazing work :fire:

So my only question is how can we get this into CI? I see there are some services that run fuzz testing against master non-stop, should we sign up for something like that?
